### PR TITLE
Add raw statsd reporter

### LINKED ISSUE #327: Statsd output is not good
Aside from the actual broken protocol packets in #207 the implementation of instrumentation with statsd in this project is just really bad. Exporting gauges of pre-aggregated metrics is a serious anti-pattern to typical/normal statsd instrumentation best practices. I'm curious if there is some specific reason why this model was adopted. Is there any appetite for re-instrumenting in a more standard manner?

### PR DESCRIPTION
**Pull request by magiconair**
This patch adds a statsd reporter without aggregation by using the https://github.com/alexcesaro/statsd library.

Fixes #327

### CODE REVIEW

**Comment by magiconair** [comment]
@drawks I've addressed your comments but I'm unsure whether I should do `prefix.name` or `prefix--name` as in the other statsd impl. The [spec](https://github.com/b/statsd_spec} isn't clear on names and prefixes. Is `--` a convention to separate prefixes from names?

**Comment by drawks** [comment]
the `--` is not used in any statsd implementation I've touched. There is no notion of "prefixes" really in statsd. Some statsd client implementations allow for prepending an arbitrary string to help with namespacing metrics, but `--` is another strange quirk that seems specific to fabio...

**Comment by magiconair** [comment]
This got pulled in via the statsd lib so I'm happy to change it. Maybe it is best not to add a separator at all then you can choose what you want.

**Comment by drawks** [comment]
I'd just use the Prefix convenience function in the library you've chosen 

https://godoc.org/gopkg.in/alexcesaro/statsd.v2#Prefix

It looks like it uses a "." separator, so you'd have to special case when prefix is `""` to not pass the empty string.

**Comment by drawks** [comment]
I've just issue a PR against your branch `issue-327-raw-statsd` to use the statsd client library's own prefixing routine.

I've done some testing with your branch and it seems to work and address the issues identified in #327 

The only lingering doubt I have about this branch is that the library you've chosen doesn't seem to have much active development and has some IMHO questionable UDP socket handling behavior (I.E. failing to initialize a statsd client if the destination isn't listening but subsequently handling re-establishing connection if there was a listener on startup. In my experience it is often the expectation that instrumentation shouldn't block a service from starting as it can often times be victimized by deployment/startup/name resolution changes and races.)

**Comment by magiconair** [comment]
I'll integrated #331 but I want to take the opportunity to generally refactor the metrics approach to make supporting additional backends much simpler. I'll get rid of the `Timer` and `Counter` shims and just provide functions for metrics events just like statsd and circonus. However, I need to sweep through the code to replace metrics with just names. So bear with me for a couple more days until this is done.

I'd also like to finally attack the tags issue for DataDog and Prometheus.

**Comment by drawks** [comment]
If you are actively working on adding tag support I'll pull back from my effort to do the same. I will reiterate that I /think/ the appropriate approach here is to make available all the values normally used for filling in the `metrics.name` template available in a struct stored in Route.targets and grab any other interesting client connection scoped data from the request and pass that as a `context.Context` into the `Counter`/`Timer` update functions. The template from `metrics.name` can then be passed into the `metrics.Registry` and leave it up to each metric backend to do what it wants with the 3 pieces of info.

This would allow:
* existing flat namespaced non-label metrics to operate the same
* hybrid flat+labels (DataDog) to do a combination of a templated flat name and additional labels
* name + label only (prometheus/influx/opentsdb) to have simple metric names with full (or configurable) label sets

**Comment by magiconair** [comment]
I've created #334 which is the metrics refactor code and then #335 for the raw statsd code on top of that. That looks simpler but needs testing.

I haven't started with tag support and if you have an approach then I'm more than happy to use that. Can you show a proof-of-concept code so that we have a basis for discussion?

**Comment by pvandervelde** [comment]
@magiconair What are the odds these statsd metrics changes are going to make it into a release in the near future? We would love to have the improved statsd metrics.

**Comment by magiconair** [comment]
I’ll work on it and try to get it in the next release

**Comment by magiconair** [comment]
Please see https://github.com/fabiolb/fabio/issues/211#issuecomment-375880738

File: metrics/statsd_raw.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+}
+
+type rawStatsDCounter struct {
+	c    *alstatsd.Client
+	name string
+}
+
+func (c *rawStatsDCounter) Inc(n int64) {
+	c.c.Increment(c.name)
+}
```
Comment (parasyte) [inline metrics/statsd_raw.go]: The input `n` arg is unused.

**Pull request closed without merging**